README.md
## ncov

全国新型冠状病毒，肺炎疫情增长情况打分和全国市级打分结果展示。china novel coronavirus (nCoV) city level score and visualization.

* 由于数据和算法都存在很大缺陷打分结果只供一定的参考，切勿造谣传谣信谣 

Demo 网址： http://ncov.city

数据可视化截图：

![](http://foreversmart.cc:7444/images/20202122.png)

![](http://foreversmart.cc:7444/images/20202121.png)

#### 初衷

这些天一直在看网上公开的数据，很多数据都是静态的数据，特别是到市这一级，只能查到当前确诊人数多少，
历史人数和趋势无法查看。我所在的城市据我观察已经很多天都是 7 人没有变化了。所以我在想能否根据一个城市
最近一段时间新增确诊人数的变化和总量来一定程度上去衡量一个城市疫情的严重程度。

* 当前想了一个简单的线性算法来进行打分

    score = 前一天的 score + 今天新增确诊人数 x 10 - （累计前10之内新增确诊人数的总和）
    
##### 运行

```
// 可视化
// 打开前确保 index.html、data.txt、merge.json 文件存在
在浏览器中 打开 index.html

```

```
// 更新结果数据
// 运行前保证 main.go DXYArea.csv 文件的存在
cd ncov/
go run main.go
// 运行后会生成 结果数据存到 data.txt 中
```

```cassandraql
// 自定义 Calc 函数来自定义城市的打分计算函数
// city 数据结构中可以方便的获取该城市历史确诊人数的数据点，这些数据点以数组的形式按更新时间降序排列
func (c *City) Calc() int {
  return 0
}

```

##### 数据来源

* 疫情历史数据：

> DXYArea.csv 到 DXY-2019-nCoV-Data 中获取最新数据
把 DXY-2019-nCoV-Data/csv 下的 DXYArea.csv 替换当前文件 
 
* 数据源：https://github.com/BlankerL/DXY-2019-nCoV-Data

##### 可视化

* 使用 echart 快速的进行可视化 https://www.echartsjs.com/en/index.html

# 中国加油，武汉加油！



history.txt
铜川市&01.25|3|01.26|5|01.27|5|01.28|5|01.29|5|01.30|5|01.31|5|02.01|5|02.02|6|02.03|7|02.04|7|02.05|7|02.06|7|02.07|8|02.08|8|02.09|8|02.10|8|02.11|8|02.12|8|02.13|8|02.14|8|02.15|8|02.16|8|02.17|8|02.18|8|02.19|8|02.20|8|02.21|8|02.22|8|02.23|8|02.24|8|02.25|8|02.26|8|02.27|8|02.28|8|02.29|8|03.01|8|03.02|8|03.04|8|03.05|8|03.06|8|03.08|8|03.10|8|03.11|8|03.12|8|&兵团第九师市&02.07|4|02.08|4|02.09|4|02.10|4|02.11|4|02.12|4|02.13|4|02.14|4|02.15|4|02.16|4|02.17|4|02.18|4|02.19|4|02.20|4|02.21|4|02.22|4|02.23|4|02.24|4|02.26|4|02.27|4|02.28|4|02.29|4|03.01|4|03.02|4|03.03|4|03.04|4|03.05|4|03.06|4|03.07|4|03.08|4|&丹东市&01.26|1|01.27|4|01.28|5|01.29|5|01.30|5|01.31|5|02.01|6|02.02|6|02.03|6|02.04|6|02.05|7|02.06|7|02.07|7|02.08|7|02.09|7|02.10|7|02.11|7|02.12|7|02.13|7|02.14|7|02.15|7|02.16|7|02.17|7|02.18|7|02.19|7|02.20|7|02.21|7|02.22|7|02.23|7|02.24|7|02.25|7|02.26|7|02.27|7|02.29|7|03.01|8|03.02|8|03.03|11|03.07|11|03.08|11|03.10|11|03.12|11|03.14|11|03.15|11|&贵港市&02.02|4|02.03|5|02.04|5|02.05|5|02.06|7|02.07|7|02.08|7|02.09|7|02.10|8|02.11|8|02.12|8|02.13|8|02.14|8|02.15|8|02.16|8|02.17|8|02.18|8|02.19|8|02.20|8|02.21|8|02.22|8|02.23|8|02.24|8|02.25|8|02.26|8|02.27|8|02.28|8|02.29|8|03.01|8|03.02|8|03.03|8|03.04|8|03.05|8|03.06|8|03.07|8|03.08|8|03.09|8|03.10|8|03.11|8|03.12|8|03.13|8|&海口市&01.24|3|01.25|3|01.26|3|01.27|6|01.28|7|01.29|9|01.30|9|01.31|9|02.01|11|02.02|11|02.03|13|02.04|14|02.05|16|02.06|16|02.07|21|02.08|23|02.09|23|02.10|28|02.11|29|02.12|29|02.13|29|02.14|31|02.15|33|02.16|33|02.17|34|02.18|34|02.19|34|02.20|39|02.21|39|02.22|39|02.23|39|02.24|39|02.25|39|02.26|39|02.27|39|02.28|39|02.29|39|03.01|39|03.03|39|03.04|39|03.05|39|03.08|39|03.12|39|&淮安市&01.26|1|01.27|2|01.28|2|01.29|3|01.30|7|01.31|10|02.01|12|02.02|13|02.03|18|02.04|22|02.05|26|02.06|34|02.07|37|02.08|40|02.09|44|02.10|46|02.11|46|02.12|47|02.13|51|02.14|56|02.15|60|02.16|63|02.17|65|02.18|66|02.19|66|02.20|66|02.21|66|02.22|66|02.23|66|02.24|66|02.25|66|02.26|66|02.27|66|02.28|66|02.29|66|03.01|66|03.02|66|03.03|66|03.04|66|03.05|66|03.06|66|03.07|66|03.08|66|03.09|66|03.10|66|03.12|66|03.13|66|03.15|66|&楚雄州市&02.11|4|02.12|4|02.13|4|02.14|4|02.15|4|02.16|4|02.17|4|02.18|4|02.19|4|02.20|4|02.21|4|02.22|4|02.23|4|02.24|4|02.25|4|02.26|4|02.27|4|02.28|4|02.29|4|03.01|4|03.02|4|03.03|4|03.06|4|03.15|4|&第七师市&01.28|1|01.29|1|01.30|1|01.31|1|02.01|1|02.02|1|02.03|1|02.04|1|02.05|1|02.06|1|&成都市&01.24|7|01.25|16|01.26|22|01.27|33|01.28|37|01.29|46|01.30|59|01.31|69|02.01|72|02.02|73|02.03|77|02.04|87|02.05|92|02.06|97|02.07|102|02.08|109|02.09|120|02.10|123|02.11|124|02.12|125|02.13|131|02.14|135|02.15|139|02.16|140|02.17|140|02.18|140|02.19|141|02.20|141|02.21|143|02.22|143|02.23|143|02.24|143|02.25|143|02.26|143|02.27|143|02.28|143|02.29|143|03.01|143|03.02|143|03.03|143|03.04|143|03.05|144|03.06|144|03.07|144|03.08|144|03.09|144|03.10|144|03.11|144|03.12|144|03.13|144|03.14|144|03.15|144|&遂宁市&01.24|1|01.25|2|01.26|2|01.27|2|01.28|3|01.29|3|01.30|4|01.31|5|02.01|5|02.02|5|02.03|5|02.04|5|02.05|5|02.06|5|02.07|5|02.08|6|02.09|8|02.10|10|02.11|10|02.12|10|02.13|10|02.14|10|02.15|10|02.16|12|02.17|17|02.18|17|02.19|17|02.20|17|02.21|17|02.22|17|02.23|17|02.24|17|02.25|17|02.26|17|02.27|17|02.28|17|02.29|17|03.01|17|03.02|17|03.03|17|03.04|17|03.05|17|03.06|17|03.07|17|03.08|17|03.09|17|03.10|17|03.11|17|03.12|17|03.13|17|03.14|17|03.15|17|&鞍山市&02.01|1|02.02|1|02.03|1|02.04|1|02.05|1|02.06|3|02.07|3|02.08|3|02.09|3|02.10|3|02.11|3|02.12|3|02.13|3|02.14|4|02.15|4|02.16|4|02.17|4|02.18|4|02.19|4|02.20|4|02.21|4|02.22|4|02.23|4|02.24|4|02.25|4|02.26|4|02.27|4|02.29|4|03.01|4|03.02|4|03.03|4|03.07|4|03.08|4|03.10|4|03.12|4|03.14|4|03.15|4|&德宏傣族景颇族自治州&01.26|2|01.27|2|01.28|2|01.29|2|01.30|2|01.31|3|02.01|3|02.02|4|02.03|4|02.04|5|02.05|5|02.06|5|02.07|5|02.08|5|02.09|5|02.10|5|&六盘水市&01.27|1|01.28|1|01.29|1|01.30|1|01.31|8|02.02|8|02.03|9|02.04|9|02.05|9|02.06|9|02.07|9|02.08|9|02.09|10|02.10|10|02.11|10|02.12|10|02.13|10|02.14|10|02.15|10|02.16|10|02.17|10|02.18|10|02.19|10|02.20|10|02.21|10|02.22|10|02.23|10|02.25|10|02.27|10|03.02|10|03.07|10|03.08|10|03.09|10|03.10|10|03.11|10|03.12|10|03.13|10|03.14|10|&赣州市&01.25|2|01.26|7|01.27|7|01.28|10|01.29|16|01.30|21|01.31|27|02.01|30|02.02|34|02.03|35|02.04|42|02.05|48|02.06|52|02.07|59|02.08|64|02.09|69|02.10|71|02.11|72|02.12|72|02.13|76|02.14|76|02.15|76|02.16|76|02.17|76|02.18|76|02.19|76|02.20|76|02.21|76|02.22|76|02.23|76|02.24|76|02.25|76|02.26|76|02.27|76|02.28|76|02.29|76|03.01|76|03.02|76|03.03|76|03.04|76|03.05|76|03.06|76|03.07|76|03.08|76|03.09|76|03.10|76|03.11|76|03.12|76|&萍乡市&01.24|1|01.25|1|01.26|2|01.27|2|01.28|2|01.29|2|01.30|5|01.31|8|02.01|9|02.02|15|02.03|18|02.04|20|02.05|22|02.06|23|02.07|24|02.08|24|02.09|24|02.10|25|02.11|26|02.12|29|02.13|29|02.14|31|02.15|32|02.16|32|02.17|33|02.18|33|02.19|33|02.20|33|02.21|33|02.22|33|02.23|33|02.24|33|02.25|33|02.26|33|02.27|33|02.28|33|02.29|33|03.01|33|03.02|33|03.03|33|03.04|33|03.05|33|03.06|33|03.07|33|03.08|33|03.09|33|03.10|33|03.11|33|03.12|33|&淮南市&01.26|1|01.27|2|01.28|2|01.29|2|01.30|6|01.31|6|02.01|6|02.02|6|02.03|7|02.04|8|02.05|12|02.06|12|02.07|14|02.08|21|02.09|22|02.10|23|02.11|24|02.12|24|02.13|23|02.14|23|02.15|23|02.16|24|02.17|25|02.18|27|02.19|27|02.20|27|02.21|27|02.22|27|02.23|27|02.24|27|02.25|27|02.26|27|02.27|27|02.28|27|02.29|27|03.01|27|03.02|27|03.03|27|03.04|27|03.05|27|03.06|27|03.08|27|&池州市&01.24|1|01.25|1|01.26|1|01.27|1|01.28|1|01.29|1|01.30|2|01.31|4|02.01|4|02.02|6|02.03|6|02.04|10|02.05|11|02.06|11|02.07|11|02.08|11|02.09|12|02.10|15|02.11|15|02.12|16|02.13|16|02.14|17|02.15|17|02.16|17|02.17|17|02.18|17|02.19|17|02.20|17|02.21|17|02.22|17|02.23|17|02.24|17|02.25|17|02.26|17|02.27|17|02.28|17|02.29|17|03.01|17|03.02|17|03.03|17|03.04|17|03.05|17|03.06|17|03.08|17|&周口市&01.24|1|01.25|4|01.26|5|01.27|11|01.28|15|01.29|19|01.30|25|01.31|36|02.01|38|02.02|40|02.03|47|02.04|52|02.05|56|02.06|59|02.07|60|02.08|62|02.09|62|02.10|65|02.11|65|02.12|66|02.13|68|02.14|69|02.15|70|02.16|71|02.1

main.go
package main

import (
	"bytes"
	"encoding/csv"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

type City struct {
	Province string  `json:"province"`
	City     string  `json:"city"`
	Data     []*Data `json:"data"`
}

type Province struct {
	Province string  `json:"province"`
	Data     []*Data `json:"data"`
}

type Data struct {
	Confirmed   int       `json:"confirmed"`
	Suspected   int       `json:"suspected"`
	Cured       int       `json:"cured"`
	Dead        int       `json:"dead"`
	UpdatedTime time.Time `json:"updated_time"`
}

const timeFormat = "2006-01-02 15:04:05.000"

func main() {
	pwd, _ := os.Getwd()
	filename := filepath.Join(pwd, "DXYArea.csv")

	file, err := os.OpenFile(filename, os.O_RDONLY, os.ModePerm)
	if err != nil {
		println("open file", filename, err)
		return
	}

	r := csv.NewReader(file)
	records, err := r.ReadAll()
	if err != nil {
		println("read csv data", err)
		return
	}

	cityMap := make(map[string]*City)
	provinceMap := make(map[string]*Province)
	for i, line := range records {
		if i == 0 {
			continue
		}

		var (
			provinceName, cityName                                                                string
			provinceConfirmedCount, provinceSuspectedCount, provinceCuredCount, provinceDeadCount string
			cityConfirmedCount, citySuspectedCount, cityCuredCount, cityDeadCount                 string
			updateTime                                                                            string
		)

		var (
			provinceConfirmed, provinceSuspected, provinceCured, provinceDead int64
			cityConfirmed, citySuspected, cityCured, cityDead                 int64
		)
		for j, item := range line {
			switch j {
			case 0:
				provinceName = item
			case 3:
				cityName = item
			case 6:
				provinceConfirmedCount = item
				provinceConfirmed, _ = strconv.ParseInt(provinceConfirmedCount, 10, 64)
			case 7:
				provinceSuspectedCount = item
				provinceSuspected, _ = strconv.ParseInt(provinceSuspectedCount, 10, 64)
			case 8:
				provinceCuredCount = item
				provinceCured, _ = strconv.ParseInt(provinceCuredCount, 10, 64)
			case 9:
				provinceDeadCount = item
				provinceDead, _ = strconv.ParseInt(provinceDeadCount, 10, 64)
			case 10:
				cityConfirmedCount = item
				cityConfirmed, _ = strconv.ParseInt(cityConfirmedCount, 10, 64)
			case 11:
				citySuspectedCount = item
				citySuspected, _ = strconv.ParseInt(citySuspectedCount, 10, 64)
			case 12:
				cityCuredCount = item
				cityCured, _ = strconv.ParseInt(cityCuredCount, 10, 64)
			case 13:
				cityDeadCount = item
				cityDead, _ = strconv.ParseInt(cityDeadCount, 10, 64)
			case 14:
				updateTime = item

			}
		}

		updated, _ := time.ParseInLocation(timeFormat, updateTime, time.FixedZone("UTC", 8*60*60))

		cityName = GetCityName(cityName, provinceName)
		provinceData := &Data{
			Confirmed:   int(provinceConfirmed),
			Suspected:   int(provinceSuspected),
			Dead:        int(provinceDead),
			Cured:       int(provinceCured),
			UpdatedTime: updated,
		}

		var cityData *Data
		if strings.HasSuffix(provinceName, "市") {
			cityName = provinceName
			cityData = provinceData

		} else {
			cityData = &Data{
				Confirmed:   int(cityConfirmed),
				Suspected:   int(citySuspected),
				Dead:        int(cityDead),
				Cured:       int(cityCured),
				UpdatedTime: updated,
			}

		}

		if _, ok := cityMap[cityName]; !ok {
			cityMap[cityName] = &City{
				Province: provinceName,
				City:     cityName,
				Data:     make([]*Data, 0, 10),
			}
		}

		cityMap[cityName].Data = append(cityMap[cityName].Data, cityData)

		if _, ok := provinceMap[provinceName]; !ok {
			provinceMap[provinceName] = &Province{
				Province: provinceName,
				Data:     make([]*Data, 0, 10),
			}
		}

		provinceMap[provinceName].Data = append(provinceMap[provinceName].Data, provinceData)
	}

	buf := &bytes.Buffer{}
	resarr := make([]int, 0, 10)
	for city, v := range cityMap {
		res := v.Calc()
		resarr = append(resarr, res)

		buf.WriteString(fmt.Sprintf("%s&%d&", city, res))
	}

	sort.Ints(resarr)
	l := len(resarr)
	fmt.Println("res city length", l)
	fmt.Println("city score rank：")
	fmt.Println(resarr[int(float32(l)*0.3)])
	fmt.Println(resarr[int(float32(l)*0.5)])
	fmt.Println(resarr[int(float32(l)*0.65)])
	fmt.Println(resarr[int(float32(l)*0.75)])
	fmt.Println(resarr[int(float32(l)*0.80)])
	fmt.Println(resarr[int(float32(l)*0.85)])
	fmt.Println(resarr[int(float32(l)*0.90)])
	fmt.Println(resarr[int(float32(l)*0.94)])
	fmt.Println(resarr[int(float32(l)*0.98)])
	fmt.Println(resarr[int(float32(l)*0.99)])

	err = ioutil.WriteFile(filepath.Join(pwd, "data.txt"), buf.Bytes(), os.ModePerm)
	if err != nil {
		fmt.Println(err)
	}

	// generate city history file
	buf.Reset()
	for city, v := range cityMap {
		history := v.History()
		buf.WriteString(city)
		buf.WriteString("&")
		for i := len(history) - 1; i >= 0; i-- {
			buf.WriteString(fmt.Sprintf("%s|%d|", history[i].UpdatedTime.Format("01.02"), history[i].Confirmed))
		}
		buf.WriteString("&")
	}

	err = ioutil.WriteFile(filepath.Join(pwd, "history.txt"), buf.Bytes(), os.ModePerm)
	if err != nil {
		fmt.Println(err)
	}

}

const score = 10

func (c *City) Calc() int {
	day := -1
	arr := make([]*Data, 0, 5)
	for _, d := range c.Data {
		if day == d.UpdatedTime.Day() {
			continue
		}

		day = d.UpdatedTime.Day()

		arr = append(arr, d)
	}

	count := 0
	for i := len(arr) - 1; i >= 0; i-- {
		if i == len(arr)-1 {
			count = arr[i].Confirmed * score
			continue
		}

		delta := arr[i].Confirmed - arr[i+1].Confirmed
		count = count + delta*score
		for j := i + 1; j <= len(arr)-1; j++ {
			if j-i > 10 {
				break
			}

			if j < len(arr)-1 {
				count = count - (arr[j].Confirmed - arr[j+1].Confirmed)
			} else {
				// j == len(arr) -1
				count = count - arr[j].Confirmed
			}

		}

		// check bonus
		if i < len(arr)-2 {
			d1 := arr[i].Confirmed - arr[i+1].Confirmed
			d2 := arr[i+1].Confirmed - arr[i+2].Confirmed

			count = count - d2 + d1
			if d1 == d2 {
				count = count - 1
			}
		}

	}

	if count < 0 {
		return 0
	}

	return count

}

func (c *City) History() (history []*Data) {
	day := -1
	arr := make([]*Data, 0, 5)
	for _, d := range c.Data {
		if day == d.UpdatedTime.Day() {
			continue
		}

		day = d.UpdatedTime.Day()

		arr = append(arr, d)
	}

	return arr
}

func GetCityName(cityName, provinceName string) (res string) {
	switch cityName {
	case "神农架林区":
		return cityName
	case "恩施", "恩施州":
		return "恩施土家族苗族自治州"
	case "湘西自治州":
		return "湘西土家族苗族自治州"
	case "大兴安岭":
		return "大兴安岭地区"
	case "黔东南州":
		return "黔东南苗族侗族自治州"
	case "黔西南州":
		return "黔西南布依族苗族自治州"
	case "黔南州":
		return "黔南布依族苗族自治州"
	case "兴安盟乌兰浩特":
		return "兴安盟"
	case "阿坝州":
		return "阿坝藏族羌族自治州"
	case "甘孜州":
		return "甘孜藏族自治州"
	case "凉山州":
		return "凉山彝族自治州"
	case "西双版纳":
		return "西双版纳傣族自治州"
	case "德宏":
		return "德宏傣族景颇族自治州"
	case "大理":
		return "大理白族自治州"
	case "红河":
		return "红河哈尼族彝族自治州"
	case "伊犁州":
		return "伊犁哈萨克自治州"
	case "阿克苏":
		return "阿克苏地区"
	case "文山":
		return "文山壮族苗族自治州"
	case "楚雄":
		return "楚雄彝族自治州"
	case "琼中县":
		return "琼中黎族苗族自治县"
	case "定安":
		return "定安县"
	case "陵水县":
		return "陵水黎族自治县"
	case "昌江":
		return "昌江黎族自治县"
	case "乐东":
		return "乐东黎族自治县"
	case "临夏":
		return "临夏回族自治州"
	case "甘南":
		return "甘南藏族自治州"
	}

	if strings.HasSuffix(cityName, "县") {
		return cityName
	}

	if strings.Contains(cityName, "锡林郭勒盟") {
		return "锡林郭勒盟"
	}

	if !strings.HasSuffix(provinceName, "市") && !strings.HasSuffix(cityName, "自治州") && !strings.HasSuffix(cityName, "市") {
		res = cityName + "市"
	} else {
		res = cityName
	}

	if strings.Contains(res, "(") {
		items := strings.Split(res, "(")
		aItems := strings.Split(items[1], ")")
		res = items[0] + aItems[1]
		return
	}

	if strings.Contains(res, "（") {
		items := strings.Split(res, "（")
		aItems := strings.Split(items[1], "）")
		res = items[0] + aItems[1]
		return
	}

	return
}
